### PR TITLE
Adds automated category-loading, and some other changes..

### PR DESCRIPTION
**Pull request by AnthyG**
**WARNING**
Dear @krixano, please create a backup of `content.json` and `data/data.json`!
Also, you have to change your the name of the property `category_name` in `categories` of your `data/data.json`-File to `name`, as it is self-explaining that the name is meant to be the categories name.

I furthermore recommend doing a backup of the complete Zite, just in case anything goes wrong!

Now to the fun part:
The major change coming with this pull-request is the automated loading of the categories, placed in the `data/data.json`-File.
This resulted in the removal of many lines and thus, much cleaner code.

I also don't want to guarantee the functionality of this code, as I can't verify that it actually works.

### CODE REVIEW

**Comment by AnthyG** [comment]
Should I also merge the [search-feature-Branch](https://github.com/AnthyG/ImportantZites/tree/search-feature) of mine into my master?
Or shall I just do a new pull-request for that ❔

**Comment by krixano** [comment]
Maybe do a different pull request for that - I'm not sure. I've never collaborated on any projects before. I would say a different one so that we keep pull requests for different features separate. What do you think is best?

**Comment by AnthyG** [comment]
Yes, I agree with that 👍

**Comment by AnthyG** [comment]
I would then recommend, to create a backup of your Zites directory, as I deleted some important files like the Zites `content.json`.

I will create another pull-request, when you merge this one.

**Comment by krixano** [comment]
I'm currently seeing a bug where the zites don't load until I sign the content.json file - but when I refresh, I have to resign it to load the zites again. There also seems to be a problem with the tabClick function. I am looking into these things right now.

**Comment by krixano** [comment]
I'm getting an error on the "this.cmd()" function call in the loadZites() function saying that "this" is undefined.

**Comment by AnthyG** [comment]
I'll fix that very quick (the "this"-problem).

**Comment by AnthyG** [comment]
I don't know, what the bug with the Zites not loading is..
But I also have that problem with the tabClick-Function, resulting in entries not getting hidden, when a tab is clicked 😞

File: index.html [hunk 1]
```diff
@@ -1,377 +1,317 @@
...
+            }*/
+
+            onOpenWebsocket() {
+                //this.messages = document.getElementById("messages")
+                //this.addMessage("System", "Ready to call ZeroFrame API!")
+                this.cmd("siteInfo", {}, (site_info) => {
+                    if (site_info.cert_user_id)
+                        document.getElementById("select_user").innerText = site_info.cert_user_id
+                    this.site_info = site_info
+                    this.loadCategories(page.loadZites);
```
Comment (krixano) [inline index.html]: I think `page.loadZites` should be `this.loadZites`?

Comment (AnthyG) [inline index.html]: No, `page.loadZites` is correct.
I forgot some `this`'s

**Comment by krixano** [comment]
How would I commit changes on this pull request? Is that allowed? I haven't used pull requests all that much before.

File: index.html (hunk 1 above)
Comment (krixano) [inline index.html]: Oh, ok. I'm not used to using functions as parameters.

**Comment by AnthyG** [comment]
I honestly don't know.. 🤔
You'd probably either have to edit it after merging, or fork my fork, edit it, do a pull-request, I accept it, and it appears here.. 😆

I'm not an git-expert either, so I can't really tell...

**Comment by AnthyG** [comment]
Oh, it seems, I deleted an if-statement that I didn't replace..: https://github.com/krixano/ImportantZites/blob/master/index.html#L258

File: index.html [hunk 2]
```diff
@@ -1,377 +1,317 @@
...
+                TweenLite.set(cats_indicator, {
+                    autoAlpha: 0
+                });
+
+                page.getCategoriesList(function(cats) {
+                    cats_indicator.innerHTML = cats.length + " categories"
+
+                    for (var x in cats) {
+                        var cat = cats[x]
+                        cats_list.innerHTML += '<li><a onclick="tabClick(' + cat.name + ', this); return false;">' + cat.name.toUpperCase() + '</a></li>'
```
Comment (krixano) [inline index.html]: This is the problem with the tabs. Quotes are not getting put around the category name in the `tabClick` function.

Comment (AnthyG) [inline index.html]: ......... 😮
# Wow

**Comment by AnthyG** [comment]
Any more issues/ errors/ *mistakes*?

**Comment by AnthyG** [comment]
The entries don't seem to get hidden, when another tab is selected 💢

**Comment by krixano** [comment]
Now I'm getting an `unexpected token }` error, but it doesn't tell me the line it's on

**Comment by AnthyG** [comment]
Do you get it, when you just load the Zite?
Or does it happen, when you do a specific action :?

**Comment by krixano** [comment]
When clicking on a tab

**Comment by AnthyG** [comment]
I'll just add `;` everywhere..
Not expecting it to be fixed though.

**Comment by AnthyG** [comment]
Well that took some time, I don't really know why..
Also fixed a little thing, where I forgot to escape the `'`

File: index.html [hunk 3]
```diff
@@ -1,377 +1,325 @@
...
+            if (id == "all") {
+                // Show all divs except the previousId (because that one is already shown)
+                page.getCategoriesList(function(cats) {
+                    for (var x in cats) {
+                        var cat = cats[x];
+
+                        if (previousId != cat.name) {
+                            var s = document.querySelectorAll('.cat-' + cat.name);
+                            for (var x2 in s) {
+                                s[x2].style.display = "";
```
Comment (krixano) [inline index.html]: I'm getting an error here saying `cannot set property display of undefined`.

File: index.html [hunk 4]
```diff
@@ -1,377 +1,325 @@
...
+                // }
+            } else {
+                page.getCategoriesList(function(cats) {
+                    for (var x in cats) {
+                        var cat = cats[x];
+
+                        if (id != cat.name) {
+                            var s = document.querySelectorAll('.cat-' + cat.name);
+                            for (var x2 in s) {
+                                s[x2].style.display = "none";
```
Comment (krixano) [inline index.html]: I'm getting an error here saying cannot set property display of undefined. Not sure why, though.

Comment (krixano) [inline index.html]: I found out that it is an index out of bounds error. The x2 is going past the last element's index in the array.

Comment (AnthyG) [inline index.html]: If I'm correct, this is fixed on the search-feature-branch..

Comment (krixano) [inline index.html]: Switching it back to this for loop fixes it:
```
for (var i = 0; i < s.length; i++) {
    s[i].style.display = "none";
}
```

Comment (AnthyG) [inline index.html]: I always used an if-statement :sweat_smile:, though that area is not fixed as I thought..

Comment (krixano) [inline index.html]: I believe the same thing is happening to the for loop on line 259 also.

Comment (AnthyG) [inline index.html]: Fixed, I hope (am on my mobile phone now.. x))

**Comment by krixano** [comment]
Two more issues:
1. Clicking on a tab doesn't set the previous one as not active
2. After a refresh, clicking on a tab works correctly, *however*, clicking on another tab doesn't.

**Comment by AnthyG** [comment]
I'll look into this.. today.. but like in 14+h :laughing: (or earlier, if I can)

**Comment by krixano** [comment]
I found out how I can also commit changes on this pull request. Basically, I was able to commit and push directly to your fork because you forked my repository. You can find out more about it here: https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by krixano** [comment]
So now, one last bug to solve - the "All" tab is never set as not active.

**Comment by AnthyG** [comment]
Today I won't be much at home, but as far as I checked the code from my phone, everything should be fine, so I don't know what keeps the "All"-Tab from not hiding >:/

**Comment by AnthyG** [comment]
When doing this (I have to do it, as I'm not able to clone the Zite), tab-changing works just fine 🤔
```
getCategoriesList(cb) {
    if (typeof cb === "function")
        cb([{
            "name": "info"
        }])

    // page.cmd("dbQuery", ["SELECT * FROM categories"], (cats) => {
    //     if (typeof cb === "function")
    //         cb(cats);
    // })
}
```

~~Have you changed the `data/data.json`-Files category-property `category_name` to `name`, and reloaded and rebuilt your Database?~~
... Has to be, as you can obviously see the category-tabs 😅

**Comment by krixano** [comment]
Another thing, redoing a database query for each tab click to get the categories seems inefficient (and will lag older computer's), so we should cache the categories in a variable and in the getCategories function, check if the variable is not null and return that, or do the database query and store the categories in the variable so that next time the function is called, the variable will not be null (and will be directly returned without the query). What do you think?

File: index.html [hunk 5]
```diff
@@ -1,377 +1,347 @@
...
+                        autoAlpha: 2
+                    });
+                })
+            }
+        }
+
+        page = new ZeroChat();
+
+        var previousId = "all";
+        var previousActiveTab = document.getElementById('allTab');
```
Comment (krixano) [inline index.html]: It seems like this might not be getting the right element.

**Comment by AnthyG** [comment]
Maybe select it on readystate of the document?
Either `document.onready = function...`, or a self-executing function `(function(){...})`

**Comment by krixano** [comment]
No, that didn't work. Two things I noticed:
1. Putting `toggleClass(document.getElementById('allTab'), "is-active")` in the `onOpenWebsocket` function right before the categories are loaded does toggle the class on the allTab element.
2. However, putting `var previousActiveTab = document.getElementById('allTab');` in the `onOpenWebsocket` function right before the `loadCategories` function call doesn't fix anything.

*I think* this means that is has something to do with the variable itself, and*not* the `document.getElementById('allTab')` part.

**Comment by AnthyG** [comment]
Then maybe triggering a click-event on the all tab fixes it?

**Comment by krixano** [comment]
That works. There is just one small thing, though it isn't that big of a deal, which is that it takes like a second for the tab to highlight/select.

**Comment by AnthyG** [comment]
@krixano one can't select a tab twice anyways.. so it's technically fixed :laughing:

**Comment by krixano** [comment]
Okay, so I think I am ready to merge this now.

**Comment by AnthyG** [comment]
Yay :tada::smile:

**Pull request merged**